File: Semestrul 3/MAP/lab 2/src/Factory/TaskContainerFactorySingleton.java
package Factory;

import Containers.Container;
import Containers.QueueContainer;
import Containers.StackContainer;
// TODO 7
public class TaskContainerFactorySingleton implements Factory{
    private static TaskContainerFactorySingleton instance;
    private QueueContainer queue;
    private StackContainer stack;

    private TaskContainerFactorySingleton() {}
    static public TaskContainerFactorySingleton getInstance(){
        if(instance == null){
            instance = new TaskContainerFactorySingleton();
        }
        return instance;
    }
    @Override
    public Container createContainer(Strategy strategy) {
        switch (strategy){
            case FIFO -> {
                if(queue == null){
                    queue = new QueueContainer();
                }
                return queue;
            }
            case LIFO -> {
                if(stack == null){
                    stack = new StackContainer();
                }
                return stack;
            }
        }
        return null;
    }
}

File: Semestrul 3/MAP/lab 10/factory/TaskContainerFactorySingleton.cs
namespace lab10.factory;

public class TaskContainerFactorySingleton : IFactory
{
    private static TaskContainerFactorySingleton? _instance;
    private QueueContainer? _queue;
    private StackContainer? _stack;

    private TaskContainerFactorySingleton()
    {
    }

    public static TaskContainerFactorySingleton GetInstance()
    {
        return _instance ??= new TaskContainerFactorySingleton();
    }

    public IContainer? CreateContainer(Strategy strategy)
    {
        return strategy switch
        {
            Strategy.Fifo => _queue ??= new QueueContainer(),
            Strategy.Lifo => _stack ??= new StackContainer(),
            _ => null
        };
    }
}